cpp/ql/test/library-tests/rangeanalysis/SimpleRangeAnalysis/test.c

```diff
@@ -463,3 +463,30 @@ int test_unsigned_mult02(unsigned b) {
 
   return total;
 }
+
+unsigned long mult_rounding() {
+  unsigned long x, y, xy;
+  x = y = 1000000003UL; // 1e9 + 3
+  xy = x * y;
+  return xy; // BUG: upper bound should be >= 1000000006000000009UL
```

Comment:
Is this a floating point accuracy issue?

Does it reflect any real world issues you've seen?

Could you explain the other bugs in this file as well (assuming you know what's going on with them) and whether you think it's going to be feasible to fix them.

Reply:
I haven't seen any real-world issues yet, but maybe CPP-Differences will uncover issues.

When I wrote these tests, I was looking for issues of the sort that would make `PointlessComparison.ql` claim that `x * x == x_squared` was always false because `x * x` would be carried out in floating point and misapproximated. It's inevitable that we'll misapproximate int64 calculations when they're done with float64, but we'll still be sound as long as the lower bound rounds down and the upper bound rounds up. That's how I painstakingly got things working for addition, but I don't know how to do it for multiplication.

I fear that any issues with this test are muddied by inaccurate pretty-printing of floats in the QL engine. The last three digits always seem to be 0, which makes me suspicious.

Reply:
What different between addition and multiplication?

Reply:
[This one weird trick](https://github.com/github/codeql/blob/66541f260bca41de238ede8d944437504268877e/cpp/ql/src/semmle/code/cpp/rangeanalysis/SimpleRangeAnalysis.qll#L404-L406) worked for addition, but it's not clear to me yet whether it can be ported to multiplication. Just changing `+` to `*` and `-` to `/` leads to division by zero in some cases. Before digging further into doing proper rounding, I sent out to produce and observe the problems I was trying to fix. Real-world cases would be ideal.